File: app/src/main/java/com/gmail/wondergab12/bank/repository/database/AtmsDao.java
package com.gmail.wondergab12.bank.repository.database;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import java.util.Calendar;
import java.util.Optional;

public class AtmsDao implements DatabaseDao {

    private static final String TABLE_NAME = "atms_table";
    private static final String COLUMN_ID = "id";
    private static final String COLUMN_DATE = "date";
    private static final String COLUMN_CITY = "city";
    private static final String COLUMN_CURRENCY = "currency";
    private static final String COLUMN_RESPONSE = "response";

    static final String SQL_CREATE_TABLE = String.format(
            "CREATE TABLE %s (" +
                    "%s INTEGER PRIMARY KEY AUTOINCREMENT, " +
                    "%s INTEGER, " +
                    "%s TEXT, " +
                    "%s TEXT, " +
                    "%s TEXT, " +
                    "UNIQUE (%s, %s) ON CONFLICT REPLACE)",
            TABLE_NAME,
            COLUMN_ID,
            COLUMN_DATE,
            COLUMN_CITY,
            COLUMN_CURRENCY,
            COLUMN_RESPONSE,
            COLUMN_CITY, COLUMN_CURRENCY);

    static final String SQL_DROP_TABLE = "DROP TABLE " + TABLE_NAME;

    private SQLiteDatabase db;

    AtmsDao(SQLiteDatabase db) {
        this.db = db;
    }

    @Override
    public long createRecord(String city, String currency, String response) {
        ContentValues contentValues = new ContentValues();
        long date = Calendar.getInstance().getTimeInMillis();
        contentValues.put(COLUMN_DATE, date);
        contentValues.put(COLUMN_CITY, city);
        contentValues.put(COLUMN_CURRENCY, currency);
        contentValues.put(COLUMN_RESPONSE, response);

        return db.insert(TABLE_NAME, null, contentValues);
    }

    @Override
    public Optional<DatabaseResponse> readRecord(String city, String currency) {
        Optional<DatabaseResponse> response = Optional.empty();

        Cursor cursor = db.query(TABLE_NAME, new String[]{COLUMN_DATE, COLUMN_RESPONSE},
                String.format("%s = ? AND %s = ?", COLUMN_CITY, COLUMN_CURRENCY),
                new String[]{city, currency}, null, null, null);

        if (cursor != null) {
            if (cursor.moveToFirst()) {
                int indexDate = cursor.getColumnIndex(COLUMN_DATE);
                int indexResponse = cursor.getColumnIndex(COLUMN_RESPONSE);
                response = Optional.of(new DatabaseResponse(cursor.getLong(indexDate),
                        cursor.getString(indexResponse)));
            }
            cursor.close();
        }

        return response;
    }

}

File: app/src/main/java/com/gmail/wondergab12/bank/repository/database/AtmsDao.kt
package com.gmail.wondergab12.bank.repository.database

import android.content.ContentValues
import android.database.sqlite.SQLiteDatabase
import java.lang.IllegalStateException
import java.util.*

const val TABLE_NAME = "atms_table"
const val COLUMN_ID = "id"
const val COLUMN_DATE = "date"
const val COLUMN_CITY = "city"
const val COLUMN_CURRENCY = "currency"
const val COLUMN_RESPONSE = "response"

const val SQL_CREATE_TABLE = "CREATE TABLE $TABLE_NAME (" +
        "$COLUMN_ID INTEGER PRIMARY KEY AUTOINCREMENT, " +
        "$COLUMN_DATE, INTEGER, " +
        "$COLUMN_CITY, TEXT, " +
        "$COLUMN_CURRENCY, TEXT, " +
        "$COLUMN_RESPONSE, TEXT, " +
        "UNIQUE ($COLUMN_CITY, $COLUMN_CURRENCY)"

const val SQL_DROP_TABLE = "DROP TABLE $TABLE_NAME"

class AtmsDao(private val db: SQLiteDatabase) : DatabaseDao {

    override fun createRecord(city: String?, currency: String?, response: String?): Long {
        val contentValues = ContentValues()
        val date = Calendar.getInstance().timeInMillis
        contentValues.put(COLUMN_DATE, date)
        contentValues.put(COLUMN_CITY, city)
        contentValues.put(COLUMN_CURRENCY, currency)
        contentValues.put(COLUMN_RESPONSE, response)

        return db.insert(TABLE_NAME, null, contentValues)
    }

    override fun readRecord(city: String?, currency: String?): Optional<DatabaseResponse> {
        var response = Optional.empty<DatabaseResponse>()

        val cursor = db.query(TABLE_NAME, arrayOf(COLUMN_DATE, COLUMN_RESPONSE),
                "$COLUMN_CITY = ? AND $COLUMN_CURRENCY = ?",
                arrayOf(city ?: throw WrongDatabaseParameter(),
                        currency ?: throw WrongDatabaseParameter()), null, null, null)

        if (cursor != null) {
            if (cursor.moveToFirst()) {
                val indexDate = cursor.getColumnIndex(COLUMN_DATE)
                val indexResponse = cursor.getColumnIndex(COLUMN_RESPONSE)
                response = Optional.of(DatabaseResponse(cursor.getLong(indexDate),
                        cursor.getString(indexResponse)))
            }
            cursor.close()
        }

        return response
    }

}

class WrongDatabaseParameter :
        IllegalStateException("Null parameter in database request!")